src/Agent.Worker/ResourceMetricsManager.cs

```diff
@@ -38,78 +47,329 @@ public void Setup(IExecutionContext context)
+
+                await Task.Delay(WARNING_MESSAGE_INTERVAL, _context.CancellationToken);
+            }
+        }
+
+        public async Task RunCpuUtilizationMonitor(string taskId)
+        {
+            while (!_context.CancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    var usedCpuPercentage = GetCpuInfo();
+
+                    if (usedCpuPercentage >= CPU_UTILIZATION_PERCENTAGE_THRESHOLD)
+                    {
+                        string message = $"CPU utilization is higher than {CPU_UTILIZATION_PERCENTAGE_THRESHOLD}%; currently used: {usedCpuPercentage:0.00}%";
+
+                        PublishTelemetry(message, taskId);
+
+                        break;
+                    }
+
+                }
+                catch (Exception ex)
+                {
+                    Trace.Warning($"Unable to get CPU info; {ex.Message}");
+
+                    break;
+                }
+
+                await Task.Delay(WARNING_MESSAGE_INTERVAL, _context.CancellationToken);
             }
         }
 
-        private const int c_mb = 1024 * 1024;
+        public struct DiskInfo
+        {
+            public long TotalDiskSpaceMB;
+            public long FreeDiskSpaceMB;
+            public string VolumeLabel;
+        }
+
+        public DiskInfo GetDiskInfo()
+        {
+            DiskInfo diskInfo = new();
+
+            string root = Path.GetPathRoot(System.Reflection.Assembly.GetEntryAssembly().Location);
+            var driveInfo = new DriveInfo(root);
+
+            diskInfo.TotalDiskSpaceMB = driveInfo.TotalSize / 1048576;
+            diskInfo.FreeDiskSpaceMB = driveInfo.AvailableFreeSpace / 1048576;
+
+            if (PlatformUtil.RunningOnWindows)
+            {
+                diskInfo.VolumeLabel = $"{root} {driveInfo.VolumeLabel}";
+            }
+
+            return diskInfo;
+        }
+
+        public string GetDiskInfoString()
+        {
+            try
+            {
+                var diskInfo = GetDiskInfo();
+
+                return StringUtil.Loc("ResourceMonitorDiskInfo", diskInfo.VolumeLabel, $"{diskInfo.FreeDiskSpaceMB:0.00}", $"{diskInfo.TotalDiskSpaceMB:0.00}");
+
+            }
+            catch (Exception ex)
+            {
+                return StringUtil.Loc("ResourceMonitorDiskInfoError", ex.Message);
+            }
+        }
 
         private Process _currentProcess;
 
-        public string GetCpuInfo()
+        public double GetCpuInfo()
         {
             if (_currentProcess == null)
-                return $"Unable to get CPU info";
-            try
             {
-                TimeSpan totalCpuTime = _currentProcess.TotalProcessorTime;
-                TimeSpan elapsedTime = DateTime.Now - _currentProcess.StartTime;
-                double cpuUsage = (totalCpuTime.TotalMilliseconds / elapsedTime.TotalMilliseconds) * 100.0;
+                throw new Exception("Current process is unavailable");
```

Comment:
are we expecting this to happen?

To guard, we can just use lazy initializion for _currentProcess -- or return 0

Reply:
Actually this is leftover from previous code, doesn't seems like we need additional checks here. Throw/Catch in `Setup`  method is also excessive, since the `GetCurrentProcess` (https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.getcurrentprocess?view=net-8.0) doesn't have explicit throws so I'll remove these. 
Thanks!